servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/ScanConcatPublisher.java

```diff
@@ -0,0 +1,197 @@
+                         final AsyncContextMap contextMap, final AsyncContextProvider contextProvider) {
+        original.subscribeInternal(new ScanConcatSubscriber<>(subscriber, signalOffloader.offloadSubscriber(
+                contextProvider.wrapPublisherSubscriber(subscriber, contextMap)), mapperSupplier.get()));
+    }
+
+    private static final class ScanConcatSubscriber<T, R> implements Subscriber<T> {
+        private static final long TERMINATED = Long.MIN_VALUE;
+        private static final long TERMINAL_PENDING = TERMINATED + 1;
+        /**
+         * We don't want to invoke {@link ScanConcatMapper#onError(Throwable)} for invalid demand because we may never
+         * get enough demand to deliver an {@link #onNext(Object)} to the downstream subscriber. {@code -1} to avoid
+         * {@link #demand} underflow in onNext (in case the source doesn't deliver a timely error).
+         */
+        private static final long INVALID_DEMAND = -1;
+        @SuppressWarnings("rawtypes")
+        private static final AtomicLongFieldUpdater<ScanConcatSubscriber> demandUpdater =
+                newUpdater(ScanConcatSubscriber.class, "demand");
+        private final Subscriber<? super R> subscriber;
+        private final Subscriber<? super R> offloadedSubscriber;
+        private final ScanConcatMapper<T, R> mapper;
+        private volatile long demand;
+        @Nullable
+        private Throwable errorCause;
+
+        ScanConcatSubscriber(final Subscriber<? super R> subscriber, final Subscriber<? super R> offloadedSubscriber,
+                             final ScanConcatMapper<T, R> mapper) {
+            this.subscriber = subscriber;
+            this.offloadedSubscriber = offloadedSubscriber;
+            this.mapper = requireNonNull(mapper);
+        }
+
+        @Override
+        public void onSubscribe(final PublisherSource.Subscription subscription) {
+            subscriber.onSubscribe(new PublisherSource.Subscription() {
+                @Override
+                public void request(final long n) {
+                    if (!isRequestNValid(n)) {
+                        handleInvalidDemand(n);
+                    } else if (demandUpdater.getAndAccumulate(ScanConcatSubscriber.this, n,
+                            FlowControlUtils::addWithOverflowProtectionIfNotNegative) == TERMINAL_PENDING) {
+                        demand = TERMINATED;
+                        if (errorCause != null) {
+                            deliverOnError(errorCause, offloadedSubscriber);
+                        } else {
+                            deliverOnComplete(offloadedSubscriber);
+                        }
+                    } else {
+                        subscription.request(n);
+                    }
+                }
+
+                @Override
+                public void cancel() {
+                    subscription.cancel();
```

Comment:
> After the Subscription is cancelled, additional Subscription.request(long n) MUST be NOPs.

But currently we may still do some work (deliver terminal signal) inside `request(n)`

Reply:
good point, I can prevent future deliveries.